src/types/types.zig

```diff
@@ -0,0 +1,376 @@
+
+/// Represents what the a type *is*
+///
+/// Numbers are special cased here. This means that when constraints, types
+/// like `Num(Int(Unsigned64))` should be reperesntsed as it's specific
+/// `flat_type.num` *not* as `flat_type.apply`. See 'Num' struct for additional
+/// details
+pub const Content = union(enum) {
+    const Self = @This();
+
+    flex_var: ?Ident.Idx,
+    rigid_var: Ident.Idx,
+    alias: Alias,
+    effectful,
+    pure,
+    structure: FlatType,
+    err,
+
+    // helpers //
+
+    /// Unwrap a record or return null
+    pub fn unwrapRecord(content: Self) ?Record {
+        switch (content) {
+            .structure => |flat_type| {
+                switch (flat_type) {
+                    .record => |record| {
+                        return record;
+                    },
+                    else => return null,
+                }
+            },
+            else => return null,
+        }
+    }
+
+    /// Unwrap a tag union or return null
+    pub fn unwrapTagUnion(content: Self) ?TagUnion {
+        switch (content) {
+            .structure => |flat_type| {
+                switch (flat_type) {
+                    .tag_union => |tag_union| {
+                        return tag_union;
+                    },
+                    else => return null,
+                }
+            },
+            else => return null,
+        }
+    }
+};
+
+// alias //
+
+/// A named alias to a different type
+pub const Alias = struct {
+    ident: TypeIdent,
+    args: Var.SafeList.Range,
+    backing_var: Var,
+};
+
+/// Represents an ident of a type
+/// TODO: Should this be something like CanIdent???
+pub const TypeIdent = struct {
+    const Self = @This();
+
+    ident_idx: Ident.Idx,
+    // TODO: Add module ident?
+
+    pub fn eql(store: *const Ident.Store, a: Self, b: Self) bool {
+        return Self.order(store, a, b) == .eq;
+    }
+
+    /// Get the ordering of how a compares to b
+    pub fn order(store: *const Ident.Store, a: Self, b: Self) std.math.Order {
+        const a_text = store.getText(a.ident_idx);
+        const b_text = store.getText(b.ident_idx);
+        return std.mem.order(u8, a_text, b_text);
+    }
+};
+
+// flat types //
+
+/// A "flat" data type
+pub const FlatType = union(enum) {
+    str,
+    box: Var,
+    list: Var,
+    tuple: Tuple,
+    num: Num,
+    custom_type: CustomType,
+    func: Func,
+    record: Record,
+    empty_record,
+    tag_union: TagUnion,
+    empty_tag_union,
+};
+
+// tuples //
+
+/// Represents a tuple
+pub const Tuple = struct {
+    elems: Var.SafeList.Range,
+};
+
+// numbers //
+
+/// Represents number
+///
+/// Numbers could be representsed by `type_apply` and phantom types, but
+/// that ends up being inefficient because every number type requires
+/// multiple different type entries. By special casing them here we can
+/// ensure that they have more compact representations
+pub const Num = union(enum) {
+    const Self = @This();
+
+    flex_var,
```

Comment:
In addition to `flex_var`, I this this will also need a `rigid_var: ...` (whatever type we want to use for the interned string name of the rigid var), because you can have `Num(a)` where `a` appears elsewhere in the type, e.g. in `plus : Num(a), Num(a) -> Num(a)`.